internal/wasm/sdk/opa/opa.go

```diff
@@ -0,0 +1,215 @@
+}
+
+// RemoveDataPath will update the current parsedData on the VMs by removing the value at the
+// specified path. If an error occurs the instance is still in a valid state, however
+// the data will not have been modified.
+func (o *OPA) RemoveDataPath(path []string) error {
+	return o.pool.RemoveDataPath(path)
+}
+
+// SetPolicy updates the policy for the subsequent Eval calls.
+// Returns either ErrNotReady, ErrInvalidPolicy or ErrInternal if an
+// error occurs.
+func (o *OPA) SetPolicy(p []byte) error {
+	if o.pool == nil {
+		return ErrNotReady
+	}
+
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	return o.setPolicyData(p, o.data)
+}
+
+// SetPolicyData updates both the policy and parsedData for the subsequent
+// Eval calls.  Returns either ErrNotReady, ErrInvalidPolicyOrData, or
+// ErrInternal if an error occurs.
+func (o *OPA) SetPolicyData(policy []byte, data *interface{}) error {
+	if o.pool == nil {
+		return ErrNotReady
+	}
+
+	var raw []byte
+	if data != nil {
+		var err error
+		raw, err = json.Marshal(*data)
+		if err != nil {
+			return fmt.Errorf("%v: %w", err, ErrInvalidPolicyOrData)
+		}
+	}
+
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	return o.setPolicyData(policy, raw)
+}
+
+func (o *OPA) setPolicyData(policy []byte, data []byte) error {
+	if err := o.pool.SetPolicyData(policy, data); err != nil {
+		return err
+	}
+
+	o.policy = policy
+	o.data = data
+	return nil
+}
+
+// EvalOpts define options for performing an evaluation
+type EvalOpts struct {
+	Entrypoint EntrypointID
+	Input      *interface{}
+	Metrics    metrics.Metrics
+}
+
+// Eval evaluates the policy with the given input, returning the
+// evaluation results. If no policy was configured at construction
+// time nor set after, the function returns ErrNotReady.  It returns
+// ErrInternal if any other error occurs.
+func (o *OPA) Eval(ctx context.Context, opts EvalOpts) (*Result, error) {
+	if o.pool == nil {
+		return nil, ErrNotReady
+	}
+
+	m := opts.Metrics
+	if m == nil {
+		m = metrics.New()
+	}
+
+	instance, err := o.pool.Acquire(ctx, m)
+	if err != nil {
+		return nil, err
+	}
+
+	defer o.pool.Release(instance, m)
+
+	result, err := instance.Eval(ctx, opts.Entrypoint, opts.Input, m)
+	if err != nil {
+		return nil, fmt.Errorf("%v: %w", err, ErrInternal)
+	}
+
+	return &Result{result}, nil
+}
+
+// Close waits until all the pending evaluations complete and then
+// releases all the resources allocated. Eval will return ErrClosed
+// afterwards.
+func (o *OPA) Close() {
+	if o.pool == nil {
+		return
+	}
+
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	o.pool.Close()
+}
+
+// Entrypoints returns a mapping of entrypoint name to ID for use by Eval() and EvalBool().
+func (o *OPA) Entrypoints(ctx context.Context) (map[string]EntrypointID, error) {
+	instance, err := o.pool.Acquire(ctx, metrics.New())
+	if err != nil {
+		return nil, err
+	}
+
+	defer o.pool.Release(instance, metrics.New())
+
+	return instance.Entrypoints(), nil
+}
```

Comment:
Do we need to expose this outside of the SDK? I was thinking the caller would just supply the entrypoint name and the SDK would be responsible for mapping to the internal ID (which is an implementation detail of the compiler).

Reply:
Not really, no. IIRC the intention was to allow callers to do more optimized things versus the OPA instance having to like hold onto a map of strings to look them up on each eval. As-is we more easily avoid having to stringify ref's on eval by using the ast.ValueMap to look up the IDs.

Looking into moving it all down a layer into the "OPA" instance.